src/serac/physics/materials/thermal_functional_material.hpp

```diff
@@ -0,0 +1,239 @@
+// Copyright (c) 2019-2021, Lawrence Livermore National Security, LLC and
+// other Serac Project Developers. See the top-level LICENSE file for
+// details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+/**
+ * @file thermal_functional_material.hpp
+ *
+ * @brief The material and load types for the thermal functional physics module
+ */
+
+#pragma once
+
+#include "serac/numerics/functional/functional.hpp"
+
+/// ThermalConductionFunctional helper structs
+namespace serac::Thermal {
+
+/// Linear isotropic thermal conduction material model
+struct LinearIsotropicConductor {
+  /// Density
+  double density_ = -1.0;
+
+  /// Specific heat capacity
+  double specific_heat_capacity_ = -1.0;
+
+  /// Constant isotropic thermal conductivity
+  double conductivity_ = -1.0;
+
+  /**
+   * @brief Function defining the thermal flux (constitutive response)
+   *
+   * @tparam T1 type of the temperature (e.g. tensor or dual type)
+   * @tparam T2 type of the temperature gradient (e.g. tensor or dual type)
+   * @param temperature_gradient Gradient of the temperature (du_dx)
+   * @return The thermal flux of the material model
+   */
+  template <typename T1, typename T2>
+  SERAC_HOST_DEVICE T2 operator()(T1& /* temperature */, T2& temperature_gradient) const
+  {
+    SLIC_ASSERT_MSG(conductivity_ > 0.0,
```

Comment:
A+ to helpful asserts

Reply:
Does `SLIC_ASSERT_MSG` actually work in a `__device__` context?

Reply:
That's a good question. Thoughts @white238 ?

Reply:
TIL: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#assertion

I'm not sure if this macro uses `assert(...)` or not, but that's a useful feature

https://godbolt.org/z/e4PK7e4Gz

Reply:
SLIC does not work on CUDA devices.  It's on Axom's agenda for this year.  We could toggle between the two internally though since they are just macros.